plugins/modules/ndo_l3out_interface_group_policy.py

```diff
@@ -0,0 +1,629 @@
+    template: l3out_template
+    l3out: l3out
+    name: interface_group_policy_1
+    description: Test description
+    interface_routing_policy:
+      name: ansible_l3out_interface_routing_policy
+      template: ansible_tenant_template
+    bfd:
+      enable: true
+      authentication: true
+      key_id: 1
+      key: TestKey
+    bfd_multi_hop:
+      enable: true
+      authentication: true
+      key_id: 1
+      key: TestKey
+    ospf:
+      enable: true
+      authentication_type: simple
+      key_id: 1
+      key: TestKey
+    custom_qos_policy:
+      template: ansible_tenant_template
+      name: ansible_custom_qos_policy
+    qos_priority: level1
+    state: present
+
+- name: Query an existing L3Out interface group policy with name
+  cisco.mso.ndo_l3out_interface_group_policy:
+    host: mso_host
+    username: admin
+    password: SomeSecretPassword
+    template: l3out_template
+    l3out: l3out
+    name: interface_group_policy_1
+    state: query
+  register: query_with_name
+
+- name: Query all existing L3Out interface group policies under l3out
+  cisco.mso.ndo_l3out_interface_group_policy:
+    host: mso_host
+    username: admin
+    password: SomeSecretPassword
+    template: l3out_template
+    l3out: l3out
+    state: query
+
+- name: Delete an existing L3Out interface group policy with name
+  cisco.mso.ndo_l3out_interface_group_policy:
+    host: mso_host
+    username: admin
+    password: SomeSecretPassword
+    template: l3out_template
+    l3out: l3out
+    name: interface_group_policy_1
+    state: absent
+"""
+
+RETURN = r"""
+"""
+
+import copy
+from ansible.module_utils.basic import AnsibleModule
+from ansible_collections.cisco.mso.plugins.module_utils.mso import (
+    MSOModule,
+    mso_argument_spec,
+)
+from ansible_collections.cisco.mso.plugins.module_utils.template import (
+    MSOTemplate,
+    KVPair,
+)
+from ansible_collections.cisco.mso.plugins.module_utils.utils import (
+    append_update_ops_data,
+    check_if_all_elements_are_none,
+)
+from ansible_collections.cisco.mso.plugins.module_utils.constants import QOS_LEVEL
+
+
+def main():
+    argument_spec = mso_argument_spec()
+    argument_spec.update(
+        template=dict(type="str", aliases=["l3out_template"]),
+        template_id=dict(type="str", aliases=["l3out_template_id"]),
+        l3out=dict(type="str", aliases=["l3out_name"]),
+        l3out_uuid=dict(type="str"),
+        description=dict(type="str"),
+        name=dict(type="str", aliases=["l3out_interface_group_policy"]),
+        interface_routing_policy=dict(
+            type="dict",
+            options=dict(
+                name=dict(type="str"),
+                template=dict(type="str"),
+                template_id=dict(type="str"),
+            ),
+            required_by={
+                "template": "name",
+                "template_id": "name",
+            },
+            mutually_exclusive=[
+                ["template", "template_id"],
+            ],
```

Comment:
This sub reference spec section can become a common dict in either mso.py or template.py.
We may all use the same spec for the name/template reference object parameter.
Also it can change to this instead:
```suggestion
        interface_routing_policy=dict(
            type="dict",
            options=dict(
                name=dict(type="str", required=True),
                template=dict(type="str"),
                template_id=dict(type="str"),
            ),
            required_one_of=[
                ["template", "template_id"],
            ],
            mutually_exclusive=[
                ("template", "template_id"),
            ],
```
Then we wouldn't need the `check_template_when_name_is_provided` function as the spec does the check for us.